### ISSUE TITLE
Recordbox forgets Music Folder on every startup and then becomes unresponsive

### ISSUE THREAD

**Issue by NotoriousNico** [flathub/ca.edestcroix.Recordbox#3]
I don't have an account on Codeberg, so I hope it's okay to post my issue here instead:

On first start, I select my music folder and let Recordbox scan it.
After scanning is done, everything works fine, I can play music, change settings and so on.

But after closing Recordbox and starting it again, I'm back at the first-run screen, where I have to select my Music folder again.
But after selecting it this time, Recordbox becomes unresponsive, and I have to force quit it.

I'm on Ubuntu 24.10, but also had the issue on Ubuntu 24.04.1 LTS as well.
I've tried Recordbox 0.8.2.

Please let me know if you need further information.

**Comment by edestcroix** [comment]
Posting here is fine too, thanks for letting me know about this.

If you run Recordbox from the command line does it output any errors when this happens? And if it does can you post the output? It sounds like an issue I've run into before during development where the background thread that checks and parses the music folder is crashing, which causes the UI to get stuck on the first-start screen.

**Comment by edestcroix** [comment]
Another thing I just noticed, do you have the setting "Sync on Startup" enabled? Because I just realized now while trying to debug this that when that setting specifically is disabled the first-start page never goes away.

**Comment by NotoriousNico** [comment]
Thank you for your reply, I appreciate it!

I did a fresh install of Recordbox 0.8.3 and this is the output I got on first run:

```
(Recordbox:2): Gtk-WARNING **: 10:31:18.339: Locale not supported by C library.
	Using the fallback 'C' locale.
```
In case it's important: My locale is `en_US.UTF-8`.

After the first start, I select my music library and let Recordbox scan it.
By the way, there's a small typo on this screen:
![image](https://github.com/user-attachments/assets/03a2abec-204c-407f-b73c-92df075f06c1)

It should read "_Synchronizing music_".

While Recordbox is scanning my Music, I get a bunch of these error messages in my Terminal:

```
(Recordbox:2): Gsk-WARNING **: 10:51:39.729: Clipping is broken, everything is clipped, but we didn't early-exit.
(Recordbox:2): Gsk-WARNING **: 10:51:39.730: Clipping is broken, everything is clipped, but we didn't early-exit.
(Recordbox:2): Gsk-WARNING **: 10:51:39.742: Clipping is broken, everything is clipped, but we didn't early-exit.
(Recordbox:2): Gsk-WARNING **: 10:51:39.742: Clipping is broken, everything is clipped, but we didn't early-exit.
(Recordbox:2): Gsk-WARNING **: 10:51:39.742: Clipping is broken, everything is clipped, but we didn't early-exit.
```



Anyway, after Recordbox has scanned my collection, I configure it to my liking, including these options:

**Settings - Library**

- Syn on Startup: **Enabled**
- Watch For changes: **Enabled**

![image](https://github.com/user-attachments/assets/7df575bc-0e0f-4b1a-a81c-f7fee56d859d)

In case it helps: My music library is on a second (internal) SSD.


I now exit Recordbox and start it again from Terminal. The output is exactly the same as from my first run:
```
(Recordbox:2): Gtk-WARNING **: 10:53:17.143: Locale not supported by C library.
	Using the fallback 'C' locale.
```

But now, Recordbox is showing the first-run screen again and I  get the "Recordbox isn't responding" message after around five seconds:
![image](https://github.com/user-attachments/assets/4a7990d6-9662-447d-90fe-1453cce5ddf4)

However, if I do nothing for around 20 to 30 seconds, the message goes away and I can see the regular library overview with my scanned library and all settings intact.

During all this time, the terminal output hasn't changed a bit, it's still this:
```
(Recordbox:2): Gtk-WARNING **: 10:55:57.688: Locale not supported by C library.
	Using the fallback 'C' locale.
```

Please let me know if you need more information.

**Comment by NotoriousNico** [comment]
Thank you for your reply, I appreciate it!

I did a fresh install of Recordbox 0.8.3 and this is the output I got on first run:

```
(Recordbox:2): Gtk-WARNING **: 10:31:18.339: Locale not supported by C library.
	Using the fallback 'C' locale.
```
In case it's important: My locale is `en_US.UTF-8`.

After the first start, I select my music library and let Recordbox scan it.
By the way, there's a small typo on this screen:
![image](https://github.com/user-attachments/assets/03a2abec-204c-407f-b73c-92df075f06c1)

It should read "_Synchronizing music_".

While Recordbox is scanning my Music, I get a bunch of these error messages in my Terminal:

```
(Recordbox:2): Gsk-WARNING **: 10:51:39.729: Clipping is broken, everything is clipped, but we didn't early-exit.
(Recordbox:2): Gsk-WARNING **: 10:51:39.730: Clipping is broken, everything is clipped, but we didn't early-exit.
(Recordbox:2): Gsk-WARNING **: 10:51:39.742: Clipping is broken, everything is clipped, but we didn't early-exit.
(Recordbox:2): Gsk-WARNING **: 10:51:39.742: Clipping is broken, everything is clipped, but we didn't early-exit.
(Recordbox:2): Gsk-WARNING **: 10:51:39.742: Clipping is broken, everything is clipped, but we didn't early-exit.
```



Anyway, after Recordbox has scanned my collection, I configure it to my liking, including these options:

**Settings - Library**

- Syn on Startup: **Enabled**
- Watch For changes: **Enabled**

![image](https://github.com/user-attachments/assets/7df575bc-0e0f-4b1a-a81c-f7fee56d859d)

In case it helps: My music library is on a second (internal) SSD.


I now exit Recordbox and start it again from Terminal. The output is exactly the same as from my first run:
```
(Recordbox:2): Gtk-WARNING **: 10:53:17.143: Locale not supported by C library.
	Using the fallback 'C' locale.
```

But now, Recordbox is showing the first-run screen again and I  get the "Recordbox isn't responding" message after around five seconds:
![image](https://github.com/user-attachments/assets/4a7990d6-9662-447d-90fe-1453cce5ddf4)

However, if I do nothing for around 20 to 30 seconds, the message goes away and I can see the regular library overview with my scanned library and all settings intact.

During all this time, the terminal output hasn't changed a bit, it's still this:
```
(Recordbox:2): Gtk-WARNING **: 10:55:57.688: Locale not supported by C library.
	Using the fallback 'C' locale.
```

**EDIT:**
Things get interesting, when I try to play a random track.
I first got a little toast message that says:
"Error: Resource not found."

After about two seconds, the toast goes away and a new message appears:
![image](https://github.com/user-attachments/assets/d54378e6-b6bb-44bc-b100-251ed4a41de0)
Interestingly, the lyrics to that song I try to play display just fine, as well as all other metadata of that song.
But the song itself can't be played.

However, this doesn't happen to all of my songs and I've experienced it only with one song so far.
In case it helps: The filename of the song I can't play is `1-01 - Cell #1.flac`.
So this issue might be totally unrelated, but it's still worth mentioning.


Please let me know if you need more information.

**Comment by edestcroix** [comment]
Awesome thank you for all the information, and for pointing out the typo, that's probably been there for months. 

The "Clipping is broken" errors are a bug in GTK that should be fixed soon. And the error playing that track is unrelated, but something I thought I had fixed so thank you for mentioning it. It's because of the `#` character in the filename, GStreamer really doesn't like filenames with it. I remember fixing it a while ago, but the fix must not have survived the port to Rust so I'll look into it again.

Just to confirm, Recordbox doesn't go unresponsive like that the first time it starts, and the lists are able to load like normal without anything freezing? The UI going completely unresponsive before the library appears is interesting, and rules out all my previous ideas for what's going on. I thought it was just stuck on the first-start page, not frozen.

All I can think to try right now is that you mentioned your music folder is on a second SSD, so you could be running into a limitation of the flatpak sandbox where apps running in flatpak can't request persistent access to a folder, and Recordbox is having issues when it tries to read your music folder after the first time. 

Can you try running `flatpak override ca.edestcroix.Recordbox --filesystem=path to the SSD`, or if you use Flatseal adding that path to the "Other Files" section in Recordbox's filesystem options, then reset Recordbox again? This will allow Recordbox to access the music folder without going through the sandbox.

**Comment by edestcroix** [comment]
Ok thank you. I don't think it's a problem with the sandbox then, that path in Recordbox would show up under `/run/user/...` in that case.

The only other thing I think it could be at this point is that your music library is much larger than mine which is revealing an inefficiency in the code that's retrieving the library data from Recordbox's database, although I'm not 100% confident on that because if that was the case I would think it would freeze on the first run too.

If you don't mind my asking, could you send me a copy of the database file at `~/.var/app/ca.edestcroix.Recordbox/data/Recordbox/recordbox.db`? If you put it in a zip you should be able to just attach it to your reply here it Github. It would allow me to debug this a lot easier, and I can hopefully track down what's causing the freeze.

**Comment by edestcroix** [comment]
Actually, you probably don't have to send me the database file after all, I just tested adding a lot of extra entries to the database and I think I'm able to reproduce your issue now. 

So when the database gets big enough the UI freezes while it's loading the library from the database, causing Recordbox to do what it's been doing for you. I'll look into fixing this for the next release and hopefully this is what is causing your issue.

Thank you again for reporting this, I spent a lot of time optimizing the code to read large libraries into the database, I wasn't expecting getting them back out of the database to be the bottleneck.

**Comment by NotoriousNico** [comment]
Your explanation while Recordbox is freezing makes a lot of sense.

Please let me know if you need anything else from me. Thank you for finding what's causing my issue and providing a fix in the future.

**Comment by edestcroix** [comment]
Hey, 0.9.0 should be releasing soon, and I've made some improvements that should resolve this, but feel free to reopen this issue if anything comes up.  It's definitely running faster, but I think I might still need to do another round of optimizing for the next release after this.

Again, thank you for reporting the issue and providing so much information, I hope it is resolved now.

**Comment by NotoriousNico** [comment]
Thank you so much for letting me know about the new update.
I took it for a spin, and everything runs smoothly so far:

To make sure, there was no interference with the previous version, I trashed all user data and basically started anew.

The initial scan of my music library was much, much faster. I didn't look at the time, but it was less than two minutes, I suppose. Playing files and everything else works like a charm now.

When closing Recordbox and opening it again, it now takes about 2 to 3 seconds, until the interface is fully ready and displays my whole library, which is great as well.

Thank you for all your hard work on Recordbox, I appreciate it a lot.
